File: 프로그래머스/2/1880. 땅따먹기/땅따먹기.py
def solution(land):
    answer = 0
    dp = [[0 for i in range(4)] for i in range(100001)]
    
    n = len(land)
    
    for i in range(4):
        dp[0][i] = land[0][i]
    
    for i in range(n):
        for j in range(4):
            for k in range(4):
                if j != k:
                    dp[i][j] = max(dp[i][j], land[i][j] + dp[i-1][k])
    
    for i in range(4):
        answer = max(answer, dp[n-1][i])
    
    return answer

File: 프로그래머스/2/12913. 땅따먹기/땅따먹기.js
function solution(land) {
    let dp = Array.from(Array(land.length), () => Array(4).fill(0))
    
    for(let i = 0; i<4; i++){
        dp[0][i] = land[0][i]
    }
    
    let n = land.length
    
    for(let i = 1; i<n; i++) {
        for(let j = 0; j<4; j++) {
            for(let k = 0; k<4; k++) {
                if(j !== k) {
                    dp[i][j] = Math.max(dp[i][j], land[i][j] + dp[i-1][k])
                }
            }
        }
    }

    return Math.max(...dp[n-1]);
}